Feature/add embed icon to editor toolbar

rt4914 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/main/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR focuses on adding `Embed Icon` in Markdown Toolbar. The keyboard shortcut for this icon is `CMD/CTRL+Shift+K`.
As per the issue description this PR covers following cases:


**1. If the user currently has no text selected, clicking it should insert following snippet:**

```
{% embed  %}
```

with cursor being at `{% embed {cursor-position} %}` so user can paste a URL from clipboard right away. 

**2. If the user already has text selected, clicking should insert the following snippet:**

```
{% embed text-the-user-has-selected %}
```

with `text-the-user-has-selected` still selected

**3. If the user has a whole embed syntax selected, e.g.**

```
{% embed some-url-text %}
```

Clicking the button should remove all of the embed syntax, and leave only the link text e.g.

```
some-url-text
```

The remaining text should be selected

**4. If the user has the link inside an embed syntax selected, e.g. the they have `some-url-text` selected, inside the following syntax:**

```
{% embed some-url-text %}
```

Clicking the button should remove the embed syntax, leaving only the `some-url-text` which should continue to stay selected


## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Related Issue #16780
- Closes #16780

## QA Instructions, Screenshots, Recordings

<img width="783" alt="Screenshot 2022-05-27 at 9 56 55 PM" src="https://user-images.githubusercontent.com/9396084/170740612-e40e472b-e4a7-4b4b-8fa4-56f32cc0ea50.png">

https://user-images.githubusercontent.com/9396084/173037229-1e8000fa-05f1-41b0-91b1-bda49f23b4ae.mov

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

**Note:** Will work on this next.

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [x] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

rt4914 commented:
I am facing these issues:
I have run `npx jest <path-to-file>`  locally and also `yarn test` and all tests pass in this locally but still travis-CI checks are failing. I am unable to understand the reason behind failure and also is there some command that I can run to reproduce this failure locally.
Also tried running `yarn e2e`  but was unable to run these test cases because of following error `ActiveRecord::ConnectionNotEstablished: connection to server at "::1", port 5432 failed: fe_sendauth: no password supplied` I have changed the port locally and the reason behind that it also mentioned in this PR description: https://github.com/forem/forem/pull/17729

rt4914 commented:
Just merged with latest code and all the tests passed. @aitchiss Can you please review this PR?

aitchiss commented on app/javascript/crayons/MarkdownToolbar/__tests__/markdownSyntaxFormatters.test.js:
```diff
@@ -1516,6 +1516,239 @@ describe('markdownSyntaxFormatters', () => {
+        expect(newCursorEnd).toEqual(34);
+      });
+
+      it('removes embed syntax anfd highlights the placeholder url, when placeholder url is selected and embed syntax is present', () => {
```
```suggestion
      it('removes embed syntax and highlights the placeholder url, when placeholder url is selected and embed syntax is present', () => {
```

rt4914 replied:
Thanks. committed this change directly.

aitchiss reviewed (commented):
This is all looking awesome 🔥 

The only question I have is  - if I:

- click the embed button with no selection
- I get e.g. `{% embed https://... %}`
- and then if I immediately click it again
- I get `https://...`

Do you think we should remove the placeholder `https://...` in this case? That would bring it more in line with the link formatter which adds the placeholder `url` text, but removes it again if the user clicks again to 'undo':

https://user-images.githubusercontent.com/20773163/172897989-c549d2b9-dec5-4fe5-86d5-868e528ba90c.mp4

aitchiss commented on app/javascript/crayons/MarkdownToolbar/__tests__/markdownSyntaxFormatters.test.js:
```diff
@@ -1516,6 +1516,239 @@ describe('markdownSyntaxFormatters', () => {
+        );
+      });
+
+      it('inserts embed syntax and highlights text, when selected text does not begin with http:// or https://', () => {
```
Actually - ignore this - I see the different handling of the URLs, and these tests make perfect sense!

rt4914 replied:
Actually I did thought so too that we don't need URL and non-URL tests for this feature but added it anyways because there can never be enough tests.
Nevertheless, should I update these two tests to combine into one test?

aitchiss replied:
I don't feel strongly about it, so happy to defer to whatever you think is best. I think the only downside to having the extra test is that it gives the impression to the next developer that the non-URL vs URL cases should be handled differently 🤷‍♀️

rt4914 replied:
In that case I will revert it. Thanks a lot.

rt4914 replied:
I have updated all the test cases as per your comments and @amywtlin 's comments.

rt4914 commented:
> This is all looking awesome 🔥
> 
> The only question I have is - if I:
> 
> * click the embed button with no selection
> * I get e.g. `{% embed https://... %}`
> * and then if I immediately click it again
> * I get `https://...`
> 
> Do you think we should remove the placeholder `https://...` in this case? That would bring it more in line with the link formatter which adds the placeholder `url` text, but removes it again if the user clicks again to 'undo':
> 
>  CleanShot.2022-06-09.at.17.28.34.mp4

@aitchiss This is actually a product question so maybe @amywtlin can help here.

In my opinion, we should actually remove `https://...` in such cases because I think this would be more of an "expected behaviour" as compared to keeping the "https://..." in such undo-functionality. Thoughts @amywtlin

amywtlin commented:
I think we don't need the `http://...` when inserting the syntax. Inserting  `{% embed %}` alone is good enough. 

A bonus I would love to see is having the cursor focused where the url needs to be inserted, like this:
 `{% embed {cursor position} %}`

rt4914 commented:
> I think we don't need the `http://...` when inserting the syntax. Inserting `{% embed %}` alone is good enough.
> 
> A bonus I would love to see is having the cursor focused where the url needs to be inserted, like this: `{% embed {cursor position} %}`

@amywtlin I have implemented this functionality and updated the tests. Also I have updated the PR description to reflect the changes.

Ridhwana reviewed (approved):
This looks good to me 🔥

aitchiss commented on app/javascript/crayons/MarkdownToolbar/markdownSyntaxFormatters.jsx:
```diff
@@ -714,6 +820,70 @@ export const coreSyntaxFormatters = {
+        tooltipHint: `${modifier.toUpperCase()} + SHIFT + K`,
+      };
+    },
+    getFormatting: ({ selectionStart, selectionEnd, value }) => {
```
I think now that the requirement to conditionally add `https://...` has been removed we could actually use the existing helper:

```
getFormatting: ({ selectionStart, selectionEnd, value }) =>
      undoOrAddFormattingForInlineSyntax({
        value,
        selectionStart,
        selectionEnd,
        prefix: '{% embed ',
        suffix: ' %}',
      }),
```
It would remove the need for the additional helper functions 🙂 
I tested locally, and all the test cases still passed, but see what you think!

rt4914 replied:
Thanks for simplifying this.

aitchiss reviewed (approved):
🔥 🔥 🔥
